source/client_channel_handler.c

```diff
@@ -94,11 +89,46 @@ static void s_on_time_to_ping(struct aws_channel_task *channel_task, void *arg,
         s_schedule_ping(connection);
     }
 }
-static int s_packet_handler_connack(
+
+static int s_validate_received_packet_type(
     struct aws_mqtt_client_connection_311_impl *connection,
-    struct aws_byte_cursor message_cursor) {
+    enum aws_mqtt_packet_type packet_type) {
+    { /* BEGIN CRITICAL SECTION */
+        mqtt_connection_lock_synced_data(connection);
+        /* [MQTT-3.2.0-1] The first packet sent from the Server to the Client MUST be a CONNACK Packet */
+        if (connection->synced_data.state == AWS_MQTT_CLIENT_STATE_CONNECTING &&
```

Comment:
debatable: Maybe just have the decoder check that the first packet it ever receives is a CONNACK packet? and then we don't need this validation function that every packet-handler needs to call...

I just found it weird that it grabs the lock, checks the value of `state` but doesn't change it, then the calling s_packet_handler_connack() function grabs the lock again and changes the value of `state`

I realize the decoder can't validate everything related to the context in which a packet was received, but if it can validate basic stuff ... that seems worth doing, vs spreading the complexity out here

Reply:
Didn't want to do anything other than a mechnical reworking of the packet handlers to support testing (don't need an actual connection).  Unsure if going further is a good idea from a safety standpoint.

Reply:
Also, I don't really want to put protocol features in the decoder both from a separation of concerns standpoint and the fact that it would mess up all the tests if the decoder doesn't get a connack first.

Reply:
sure, either way.
I worked around that in HTTP/2 like this:
```
    /* If true, do not expect the connection preface and immediately accept any frame type.
     * Only set this when testing the decoder itself */
    bool skip_connection_preface;
```